### PR TITLE
Lazy index

### PR DESCRIPTION
**Pull request by bidoubiwa**
# Breaking changes:

## Index is now a public Class required for most of the indexes related methods

In MeiliSearch some actions are about to the whole instance. For example `GET /stats` but most are specific actions on indexes, for exemple `GET indexes/myIndex/documents`.
All specific actions linked to an index are now available through an instance of the `Indexes` class.

### Before

```swift
client.addDocuments(documents)
```

### After

Now, these methods are only available in the `Indexes` class.

```swift
let index = client.index("myIndex")
index.getDocuments().
```

Or

```swift
client.index(myIndex").getDocuments()
```

The following method are still available in a Client instance and return an `Index` instance:

- `client.index("myIndex")`
- `createIndex(uid: "myIndex")`
- `getOrCreateIndex(uid: "myIndex")`
- `getindex("myIndex")`
- `getindexes("myIndex")` -> returns an array of Indexes()

## Breaking methods

Every method related to a specific `index` are now only available in an Index instance.

These are the methods that you will not find on a `Client` instance anymore:

- `client.index("myIndex").addDocuments`
- `client.index("myIndex").updateDocuments`
- `client.index("myIndex").updateDocuments`
- `client.index("myIndex").getDocument`
- `client.index("myIndex").getDocument`
- `client.index("myIndex").getDocuments`
- `client.index("myIndex").deleteDocument`
- `client.index("myIndex").deleteAllDocuments`
- `client.index("myIndex").deleteBatchDocuments`
- `client.index("myIndex").search`
- `client.index("myIndex").getUpdate`
- `client.index("myIndex").getAllUpdates`
- `client.index("myIndex").waitForPendingUpdate`
- `client.index("myIndex").getSettings`
- `client.index("myIndex").updateSettings`
- `client.index("myIndex").resetSettings`
- `client.index("myIndex").getSynonyms`
- `client.index("myIndex").updateSynonyms`
- `client.index("myIndex").resetSynonyms`
- `client.index("myIndex").getStopWords`
- `client.index("myIndex").updateStopWords`
- `client.index("myIndex").resetStopWords`
- `client.index("myIndex").getRankingRules`
- `client.index("myIndex").updateRankingRules`
- `client.index("myIndex").resetRankingRules`
- `client.index("myIndex").getDistinctAttribute`
- `client.index("myIndex").updateDistinctAttribute`
- `client.index("myIndex").resetDistinctAttribute`
- `client.index("myIndex").getSearchableAttributes`
- `client.index("myIndex").updateSearchableAttributes`
- `client.index("myIndex").resetSearchableAttributes`
- `client.index("myIndex").getDisplayedAttributes`
- `client.index("myIndex").updateDisplayedAttributes`
- `client.index("myIndex").resetDisplayedAttributes`
- `client.index("myIndex").getFilterableAttributes`
- `client.index("myIndex").updateFilterableAttributes`
- `client.index("myIndex").resetFilterableAttributes`
- `client.index("myIndex").getSortableAttributes`
- `client.index("myIndex").updateSortableAttributes`
- `client.index("myIndex").resetSortableAttributes`
- `client.index("myIndex").stats`

## Parameters label removed/added/renamed

Some method saw there parameter declaration changed to make the usage more explicit and to add some needed data.

Example:
`createIndex(UID: String, _ completion: ...)`
Became:
`createIndex(uid: String, primaryKey: String? = nil, _ completion: ...)`

All required `UID` labels became `uid`. As seen in the example above.

### CODE REVIEW

File: Sources/MeiliSearch/Indexes.swift [hunk 1]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+  /**
+    Add a list of documents or update them if they already exist. If the provided index does not exist, it will be created.
+
+    If you send an already existing document (same documentId) the old document will be only partially
+    updated according to the fields of the new document.
+    Thus, any fields not present in the new document are kept and remained unchanged.
+
+    To completely overwrite a document, `addDocuments`
+
+   - parameter documents:  The documents to add in MeiliSearch.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:299]:
```suggestion
   - parameter documents:  The documents to update in MeiliSearch.
```

File: Sources/MeiliSearch/Indexes.swift [hunk 2]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+   */
+  public func getDocument<T>(
+    _ identifier: Int,
+    _ completion: @escaping (Result<T, Swift.Error>) -> Void)
+  where T: Codable, T: Equatable {
+    self.documents.get(self.uid, String(identifier), completion)
+  }
+
+  /**
+   List the all Documents.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:377]:
```suggestion
   List all Documents.
```

File: Sources/MeiliSearch/Indexes.swift [hunk 3]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+
+   - parameter limit:      Limit the size of the query.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:379]: It's not limit it's an option

File: Sources/MeiliSearch/Indexes.swift [hunk 4]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+    options: GetParameters? = nil,
+    _ completion: @escaping (Result<[T], Swift.Error>) -> Void)
+  where T: Codable, T: Equatable {
+    self.documents.getAll(self.uid, options, completion)
+  }
+
+  /**
+   Delete a Document on the index based on the provided document identifier.
+
+   - parameter identifier: The document identifier for the Document to be found.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:395]: It's not identifier it's document Id

File: Sources/MeiliSearch/Indexes.swift [hunk 5]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+   */
+  public func getDocuments<T>(
+    options: GetParameters? = nil,
+    _ completion: @escaping (Result<[T], Swift.Error>) -> Void)
+  where T: Codable, T: Equatable {
+    self.documents.getAll(self.uid, options, completion)
+  }
+
+  /**
+   Delete a Document on the index based on the provided document identifier.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:393]:
```suggestion
   Delete a document on the index based on the provided document identifier.
```

Comment (alallema) [reply Sources/MeiliSearch/Indexes.swift:393]: And others

File: Sources/MeiliSearch/Indexes.swift [hunk 6]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+   */
+  public func deleteAllDocuments(
+    _ completion: @escaping (Result<Update, Swift.Error>) -> Void) {
+    self.documents.deleteAll(self.uid, completion)
+  }
+
+  /**
+   Delete a selection of documents based on array of document `identifiers`'s.
+
+   - parameter documentsIdentifiers: The array of unique identifier for the Document to be deleted.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:421]: It's documentIds

File: Sources/MeiliSearch/Indexes.swift [hunk 7]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+  where T: Codable, T: Equatable {
+    self.search.search(self.uid, searchParameters, completion)
+  }
+
+  // MARK: Updates
+
+  /**
+   Get the status of an update of the index.
+
+   - parameter update:    The update value.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:454]: It's UpdateId

File: Sources/MeiliSearch/Indexes.swift [hunk 8]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+  public func getUpdate(
+    _ updateId: Int,
+    _ completion: @escaping (Result<Update.Result, Swift.Error>) -> Void) {
+    self.updates.get(self.uid, updateId, completion)
+  }
+
+  /**
+   Get the status of an update of the index.
+
+   - parameter update:    The update value.
```
Comment (alallema) [inline Sources/MeiliSearch/Indexes.swift:468]:
```suggestion
   Get all update status of the index.

```

**Review by alallema — COMMENTED** [review]
🔥 🔥 🔥 🎉 🎉 🎉 
So great @bidoubiwa!

File: Sources/MeiliSearch/Indexes.swift [hunk 9]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+   - parameter completion: The completion closure used to notify when the server
+   completes the query request, it returns a `Result` object that contains `Update`
+   value. If the request was sucessful or `Error` if a failure occured.
+   */
+  public func resetDistinctAttribute(
+    _ completion: @escaping (Result<Update, Swift.Error>) -> Void) {
+    self.settings.resetDistinctAttribute(self.uid, completion)
+  }
+
+  // MARK: Searchable Attribute
```
Comment (curquiza) [inline Sources/MeiliSearch/Indexes.swift:699]:
```suggestion
  // MARK: Searchable Attributes
```

File: Sources/MeiliSearch/Indexes.swift [hunk 10]
```diff
@@ -119,54 +201,706 @@ struct Indexes {
...
+   - parameter completion: The completion closure used to notify when the server
+   completes the query request, it returns a `Result` object that contains `Update`
+   value. If the request was sucessful or `Error` if a failure occured.
+   */
+  public func resetSearchableAttributes(
+    _ completion: @escaping (Result<Update, Swift.Error>) -> Void) {
+    self.settings.resetSearchableAttributes(self.uid, completion)
+  }
+
+  // MARK: Displayed Attribute
```
Comment (curquiza) [inline Sources/MeiliSearch/Indexes.swift:742]:
```suggestion
  // MARK: Displayed Attributes
```

**Review by curquiza — CHANGES_REQUESTED** [review]
This kind of feature addition 👌 🎉 

Capture d’écran 2021-09-23 à 10 53 09

File: .code-samples.meilisearch.yaml [hunk 11]
```diff
@@ -751,7 +732,7 @@ settings_guide_accept_new_fields_1: |-
 add_movies_json_1: |-
   let path = Bundle.main.url(forResource: "movies", withExtension: "json")
   let documents: Data = Data(contentsOf: path)
```
[the remark this answers is not in the document]
Comment (bidoubiwa) [reply .code-samples.meilisearch.yaml:734]: Its the most accurante since it imports the movies.json file!

**Comment by curquiza** [comment]
Don't forget the `breaking-change` label?

**Review by alallema — APPROVED** [review]

**Comment by bidoubiwa** [comment]
bors merge

**Pull request merged**